include/fi_perf.h

```diff
@@ -0,0 +1,386 @@
+		perror("mmap on perf fd");
+		/* Error handling */
+		goto exit_err_2;
+	}
+
+ exit:
+	*handle_ptr = handle;
+	return result;
+
+	/* Error handling */
+ exit_err_2:
+	close(context->fd);
+ exit_err_1:
+	free(handle);
+ exit_err_0:
+	handle = NULL;
+
+	result = -1;
+	goto exit;
+}
+
+static inline void fi_perf_finalize(struct fi_perf_handle * handle)
+{
+	if (handle) {
+		struct fi_perf_rdpmc_context * context =
+			(struct fi_perf_rdpmc_context *)handle->private;
+		munmap(context->buf, sysconf(_SC_PAGESIZE));
+		close(context->fd);
+	}
+}
+
+static inline void fi_perf_reset(struct fi_perf_handle * handle, int slot_id)
+{
+	if (handle && (slot_id < handle->count)) {
+		handle->slots[slot_id].begin = 0;
+		handle->slots[slot_id].sum   = 0;
+		handle->slots[slot_id].count = 0;
+	}
+}
+
+static inline void fi_perf_reset_all(struct fi_perf_handle * handle)
+{
+	if (handle) {
+		int i;
+		for (i = 0; i < handle->count; i++) {
+			fi_perf_reset(handle, i);
+		}
+	}
+}
+
+static inline void fi_perf_set_slot_name(struct fi_perf_handle * handle,
+										 int slot_id, char * slot_name)
+{
+	if (handle && (slot_id < handle->count)) {
+		strncpy(handle->slot_names[slot_id], slot_name,
+				FI_PERF_SLOT_NAME_LENGTH_MAX - 1);
+		handle->slot_names[slot_id][FI_PERF_SLOT_NAME_LENGTH_MAX - 1] = '\0';
+
+		handle->slots[slot_id].is_active = 1; /* set name activates the slot */
+	}
+}
+
+static inline uint64_t fi_perf_rdpmc_get_value(struct fi_perf_handle * handle)
+{
+	/* There are no handle and any other checks on a critical path */
+	uint64_t value;
+	unsigned seq;
+	uint64_t offset;
+
+	struct fi_perf_rdpmc_context * context =
+		(struct fi_perf_rdpmc_context *)handle->private;
+	struct perf_event_mmap_page * buf = context->buf;
+
+	do {
+		seq = buf->lock;
+		fi_perf_rdpmc_rmb();
+#if defined(__ICC) || defined(__INTEL_COMPILER)
+		value = _rdpmc(buf->index - 1);
+#else /* GCC */
+		value = __builtin_ia32_rdpmc(buf->index - 1);
+#endif
+		offset = buf->offset;
+		fi_perf_rdpmc_rmb();
+	} while (buf->lock != seq);
+
+	return value + offset;
+}
+
+static inline void fi_perf_begin(struct fi_perf_handle * handle, int slot_id)
+{
+	/* There are no handle and any other checks on a critical path */
+	handle->slots[slot_id].begin = fi_perf_rdpmc_get_value(handle);
+}
+
+static inline void fi_perf_end(struct fi_perf_handle * handle, int slot_id)
+{
+	/* There are no handle and any other checks on a critical path */
+	/* There is no counter wrapping handling */
+	handle->slots[slot_id].sum +=
+		(fi_perf_rdpmc_get_value(handle) - handle->slots[slot_id].begin);
+	handle->slots[slot_id].count++;
+}
+
+static inline void fi_perf_dump(struct fi_perf_handle * handle,
+								int slot_id, FILE * stream)
+{
+	 if (handle &&
+		 (slot_id < handle->count) &&
+		 handle->slots[slot_id].is_active) {
+		 struct fi_perf_rdpmc_context * context =
+			 (struct fi_perf_rdpmc_context *)handle->private;
+		 fprintf(stream, "FI_PERF_RDPMC [%s] (%x, %04x) average = %g"
```

Comment:
Please use FI_LOG functions

Reply:
It looks like there is no way to have FI_LOG/INFO invocation w/o provider specification.
fi_log_enabled requires it.